forsys/pmatrix.py

```diff
@@ -67,28 +59,22 @@ def get_row(self, big_edge: object) -> Tuple:
         :return: Returns the left hand side row and the right hand side value.
         :rtype: Tuple
         """
-        total_number_cells = len(self.frame.cells)
-        lhs_row = np.zeros(total_number_cells)
-        rhs_value = 0
-
-        if len(big_edge.own_cells) > 2:
-            raise(NotImplementedError)
-        
-        c1_position = list(self.frame.cells.keys()).index(big_edge.own_cells[0])
-        c2_position = list(self.frame.cells.keys()).index(big_edge.own_cells[1])
-
-        # self.mapping_order[big_edge.own_cells[0]] = c1_position
-        # self.mapping_order[big_edge.own_cells[1]] = c2_position
+        lhs_row = np.zeros(len(self.mapping_order))  # n_cells
+        big_edge_cells = big_edge.own_cells
+        if len(big_edge_cells) != 2:
+            raise ValueError(f'big edge has own_cells={len(big_edge_cells)}, expecting 2')
 
         curvature = big_edge.calculate_total_curvature(normalized=False)
-        rhs_value = big_edge.tension * curvature    
+        rhs_value = big_edge.tension * curvature
 
-        if self.frame.cells[big_edge.own_cells[0]].get_area_sign() > 0:
+        c1_position = self.mapping_order[big_edge_cells[0]]
+        c2_position = self.mapping_order[big_edge_cells[1]]

```

Comment:
Previous code did a linear scan each time, but we can use `mapping_order` to get the index in constant time, which makes the outer loop run in linear time instead of quadratic time.